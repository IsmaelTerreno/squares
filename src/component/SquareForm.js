import React, { Component } from 'react';
import '../styles/SquareForm.css';
import { TwitterPicker } from 'react-color';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import PreviewSquare from './PreviewSquare';
import Square from '../model/Square';
import Message from '../container/Message.container';
import { SIZE_SQUARE } from '../ducks/square/square.action';

const defaultState = {
  color:"#d9e3f0",
  border: 0,
  isShowPicker: false,
  size: SIZE_SQUARE.small
};

/**
 * Form to create a Square.
 */
class SquareForm extends Component {
    static propTypes = {
      onSubmit: PropTypes.func,
      square: PropTypes.shape({
        color: PropTypes.string,
        border: PropTypes.number,
        size: PropTypes.number,
        id: PropTypes.string
      }),
      newMessage: PropTypes.func,
      onDelete: PropTypes.func
    };

    state = defaultState;

    componentDidMount(){
      if(this.props.square){
        const { color, border, size } = this.props.square;
        this.setState({
          color,
          border,
          size
        });
      }
      this.props.newMessage("","");
    }

    showColorPicker = () => {
        this.setState({
            isShowPicker: true
        });
    };

    hideColorPicker = () => {
        this.setState({
            isShowPicker: false
        });
    };

    handleChangeColor = (color) => {
        this.setState({
            color: color.hex,
            isShowPicker: true
        });
    };
    handleChangeBorder = (event) => {
      this.setState({
        border: parseInt(event.target.value, 10)
      });
    };

    handleChangeSize = (event) => {
      this.setState({
        size: event.target.value
      });
    };
    handleOnSubmit = (event) => {
      event.preventDefault();
      const { square } = this.props;
      const {color, border, size} = this.state;
      const squareId = (square) ? square.id : null;
      const squareToPersist = new Square(
        squareId,
        color,
        border,
        size
      );
      this.props.onSubmit(squareToPersist);
      const delayedAction = () => {
        this.setState(defaultState);
      };
      setTimeout(() => { delayedAction() }, 1900);
    };

    goBackTo = () => {
      const { history} = this.props;
      history.push("/square/gallery");
    };
    render(){
        const {color, isShowPicker, border, size} = this.state;
        const { square, onDelete, history } = this.props;
        return (
            <div className="animated flipInX">
                <h4>Square</h4>
                <form onSubmit={this.handleOnSubmit}>
                    <div className="form-group">
                        <label htmlFor="colourSquare">Colour</label>
                        <div className="input-group">
                            <div className="input-group-prepend" onClick={this.showColorPicker}>
                              <div className="input-group-text" >
                                <FontAwesomeIcon icon="ellipsis-v" />
                              </div>
                            </div>
                          <input id="colourSquare"
                            readOnly
                            style={{
                                backgroundColor: color,
                                color: 'rgba(0, 0, 0, 0)'
                            }}
                            value={color}
                            className="form-control"
                            placeholder="Color"
                          />
                        </div>
                        {isShowPicker &&
                            <div className="color-picker animated rubberBand">
                                <TwitterPicker className="color-picker-panel" color={ color } onChangeComplete={ this.handleChangeColor }/>
                                <button onClick={this.hideColorPicker} type="button" className="btn btn-primary color-picker-button">Ok</button>
                            </div>
                        }
                    </div>
                  {!isShowPicker &&
                    <div className="form-group">
                      <label htmlFor="sizeSquare">Size</label>
                      <select onChange={this.handleChangeSize} className="form-control" id="sizeSquare">
                        <option value={SIZE_SQUARE.small} key="size-1">Small</option>
                        <option value={SIZE_SQUARE.medium} key="size-2">Medium</option>
                        <option value={SIZE_SQUARE.big} key="size-3">Big</option>
                      </select>
                    </div>
                  }
                  {!isShowPicker &&
                  <div className="form-group">
                    <label htmlFor="borderSquare">Border</label>
                    <input min="0" max="40" value={border} onChange={this.handleChangeBorder} type="range" className="form-control-range" id="borderSquare"/>
                  </div>}
                  <h5>Preview</h5>
                  <div className="row justify-content-md-center">
                    <div className={"square-preview-message col-lg-12 col-md-12 col-sm-12 col-12"} >
                      <Message/>
                    </div>
                  </div>
                  <div className="row justify-content-md-center">
                    <PreviewSquare
                      color={color}
                      border={border}
                      size={size}
                    />
                  </div>
                  <div className="dropdown-divider space" />
                  <div className="row">
                    <div className="col">
                      <button type="button" onClick={() => {
                        onDelete(square);
                        history.push("/square/gallery");
                      }} className="btn btn-primary animated rubberBand"><FontAwesomeIcon icon="trash-alt" /></button>
                    </div>
                    <div className="col">
                      <button type="submit" className="btn btn-primary animated rubberBand"><FontAwesomeIcon icon="save" /></button>
                    </div>
                    <div className="col">
                      <button onClick={this.goBackTo} className="btn btn-primary animated rubberBand"><FontAwesomeIcon icon="images" /></button>
                    </div>
                  </div>
                </form>
            </div>
        );
    }
}

export default SquareForm;
