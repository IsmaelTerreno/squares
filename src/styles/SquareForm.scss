@import "./index.scss";

.color-picker{
  background: #f9f9f9;
  padding: 0.8rem;
}

.color-picker-panel{
  margin: 0 auto;
}

.color-picker-button{
  margin-top: 0.5rem;
}

h4{
  margin-bottom: 2rem;
}

form{
  margin-bottom: 2rem;
}

.dropdown-divider.space{
  margin-top: 2rem;
  margin-bottom: 2rem;
}

.square-preview-message{
  height: 3.5rem;
}
