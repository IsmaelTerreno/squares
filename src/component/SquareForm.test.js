import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from "enzyme-adapter-react-16";
import SquareForm from './SquareForm';

configure({ adapter: new Adapter() });

describe('SquareForm', () => {

  test('Should renders appropriately', () => {
    const mockOnSubmit = jest.fn();
    const mockNewMessage = jest.fn();
    const wrapper = shallow(
      <SquareForm
        onSubmit={mockOnSubmit}
        newMessage={mockNewMessage} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should call the function passed via onSubmit props.', () => {
    const mockOnSubmit = jest.fn();
    const mockNewMessage = jest.fn();
    const wrapper = shallow(
      <SquareForm
        onSubmit={mockOnSubmit}
        newMessage={mockNewMessage} />
    );
    const form = wrapper.find('form');
    form.simulate('submit', {
        preventDefault: () => {
        }
      }
    );
    expect(mockOnSubmit).toHaveBeenCalled();
  });

});
