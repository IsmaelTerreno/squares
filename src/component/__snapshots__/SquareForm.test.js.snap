// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SquareForm Should renders appropriately 1`] = `
<div
  className="animated flipInX"
>
  <h4>
    Square
  </h4>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="colourSquare"
      >
        Colour
      </label>
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
          onClick={[Function]}
        >
          <div
            className="input-group-text"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon="ellipsis-v"
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              symbol={false}
              title=""
              transform={null}
            />
          </div>
        </div>
        <input
          className="form-control"
          id="colourSquare"
          placeholder="Color"
          readOnly={true}
          style={
            Object {
              "backgroundColor": "#d9e3f0",
              "color": "rgba(0, 0, 0, 0)",
            }
          }
          value="#d9e3f0"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="sizeSquare"
      >
        Size
      </label>
      <select
        className="form-control"
        id="sizeSquare"
        onChange={[Function]}
      >
        <option
          key="size-1"
          value={5}
        >
          Small
        </option>
        <option
          key="size-2"
          value={8}
        >
          Medium
        </option>
        <option
          key="size-3"
          value={10}
        >
          Big
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="borderSquare"
      >
        Border
      </label>
      <input
        className="form-control-range"
        id="borderSquare"
        max="40"
        min="0"
        onChange={[Function]}
        type="range"
        value={0}
      />
    </div>
    <h5>
      Preview
    </h5>
    <div
      className="row justify-content-md-center"
    >
      <div
        className="square-preview-message col-lg-12 col-md-12 col-sm-12 col-12"
      >
        <Connect(withRouter(Message)) />
      </div>
    </div>
    <div
      className="row justify-content-md-center"
    >
      <PreviewSquare
        border={0}
        color="#d9e3f0"
        size={5}
      />
    </div>
    <div
      className="dropdown-divider space"
    />
    <div
      className="row"
    >
      <div
        className="col"
      >
        <button
          className="btn btn-primary animated rubberBand"
          onClick={[Function]}
          type="button"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="trash-alt"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </div>
      <div
        className="col"
      >
        <button
          className="btn btn-primary animated rubberBand"
          type="submit"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="save"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </div>
      <div
        className="col"
      >
        <button
          className="btn btn-primary animated rubberBand"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="images"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
